jscomp/others/node_buffer.ml

```diff
@@ -26,17 +26,208 @@
+  ?sourceStart:int ->
+  ?sourceEnd:int ->
+  unit -> int = "copy"
+[@@bs.send]
+
+external equals : t -> t -> bool = "equals"
+[@@bs.send]
+
+external fill :
+  t ->
+  ([ `String of string | `Integer of int | `Buffer of t ] [@bs.unwrap]) ->
+  ?offset:int ->
+  ?end_:int ->
+  ?encoding: ([ `ascii  | `utf8  | `utf16le  | `usc2  | `base64  | `latin1 | `binary | `hex ] [@bs.string]) ->
+  unit -> t = "fill"
+[@@bs.send]
+
+external includes :
+  t -> 
+  ?byteOffset:int ->
+  ?encoding: ([ `ascii  | `utf8  | `utf16le  | `usc2  | `base64  | `latin1 | `binary | `hex ] [@bs.string]) ->
+  unit -> bool = "includes"
+[@@bs.send]
+
+external indexOf :
+  t -> 
+  ([ `String of string | `Integer of int | `Buffer of t ] [@bs.unwrap]) ->
+  ?byteOffset:int ->
+  ?encoding: ([ `ascii  | `utf8  | `utf16le  | `usc2  | `base64  | `latin1 | `binary | `hex ] [@bs.string]) ->
+  unit -> int = "indexOf"
+[@@bs.send]
+
+external lastIndexOf :
+  t -> 
+  ([ `String of string | `Integer of int | `Buffer of t ] [@bs.unwrap]) ->
+  ?byteOffset:int ->
+  ?encoding: ([ `ascii  | `utf8  | `utf16le  | `usc2  | `base64  | `latin1 | `binary | `hex ] [@bs.string]) ->
+  unit -> int = "lastIndexOf"
+[@@bs.send]
+
+external length : t -> int = "length" 
+[@@bs.get]
+
+external readDoubleBE : t -> int -> float = "readDoubleBE" [@@bs.send]
+external readDoubleLE : t -> int -> float = "readDoubleLE" [@@bs.send]
+external readFloatBE : t -> int -> float = "readFloatBE" [@@bs.send]
+external readFloatLE : t -> int -> float = "readFloatLE" [@@bs.send]
+
+external readInt8 : t -> int -> int = "readInt8" [@@bs.send]
+external readInt16BE : t -> int -> int = "readInt16BE" [@@bs.send]
+external readInt16LE : t -> int -> int = "readInt16LE" [@@bs.send]
+external readInt32BE : t -> int -> int = "readInt32BE" [@@bs.send]
+external readInt32LE : t -> int -> int = "readInt32LE" [@@bs.send]
+external readIntBE : t -> offset:int -> byteLength:int -> int = "readIntBE" [@@bs.send]
+external readIntLE : t -> offset:int -> byteLength:int -> int = "readIntLE" [@@bs.send]
+
+external readUInt8 : t -> int -> int = "readUInt8" [@@bs.send]
+external readUInt16BE : t -> int -> int = "readUInt16BE" [@@bs.send]
+external readUInt16LE : t -> int -> int = "readUInt16LE" [@@bs.send]
+external readUInt32BE : t -> int -> int = "readUInt32BE" [@@bs.send]
+external readUInt32LE : t -> int -> int = "readUInt32LE" [@@bs.send]
+external readUIntBE : t -> offset:int -> byteLength:int -> int = "readUIntBE" [@@bs.send]
+external readUIntLE : t -> offset:int -> byteLength:int -> int = "readUIntLE" [@@bs.send]
+
+external subarray : t -> ?start: int -> ?end_: int -> unit -> t = "subarray" [@@bs.send]
+external slice : t -> ?start: int -> ?end_:int -> unit -> t = "slice" [@@bs.send]
+
+external swap16 : t -> t = "swap16" [@@bs.send]
+external swap32 : t -> t = "swap32" [@@bs.send] 
+external swap64 : t -> t = "swap64" [@@bs.send] 
+
+external toJSON : t -> Js.Json.t = "toJSON" [@@bs.send]
+
+external write :
```

Comment:
This one is a bit tricky! The [doc](https://nodejs.org/api/buffer.html#buffer_buf_write_string_offset_length_encoding) says that the `length` parameter is used only if the `offset` parameter is also used. In other words, there are two (main) forms of calling this function in JS:

```javascript
// 1
buf.write(string, offset, encoding)
buf.write(string, offset, length, encoding)

// 2
buf.write(string, encoding)
```

I think this justifies two different bindings to capture this logic. Something like:

```ocaml
external write : t -> string -> ?offset:int -> ?encoding:(...) -> unit -> int = "write" [@@bs.send]
external writeLength : t -> string -> offset:int -> ?length:int -> ?encoding:(...) -> unit -> int = "write" [@@bs.send]
```

Note how in the second binding the `offset` param is non-optional and this allows us to use the `length` param.

Reply:
Good catch! Thank you!

Reply:
Fixed